runtime/compiler/env/J9SharedCache.cpp

```diff
@@ -1567,4 +1571,108 @@ TR_J9JITServerSharedCache::storeSharedData(J9VMThread *vmThread, const char *key
+   return ptr;
+   }
+
+void *
+TR_J9DeserializerSharedCache::lookupClassLoaderAssociatedWithClassChain(void *chainData)
+   {
+   // We return chainData directly because the only thing this function can be called on is the result of
+   // TR_J9DeserializerSharedCache::pointerFromOffsetInSharedCache(), and that will return a (J9ClassLoader *)
+   // directly when given a class loader offset.
+   return chainData;
+   }
+
+bool
+TR_J9DeserializerSharedCache::classMatchesCachedVersion(J9Class *clazz, UDATA *chainData)
+   {
+   // During deserialization, we find a matching J9Class for the first entry of the chainData, meaning that
+   // its ROM class chain matches what was recorded during compilation. This provides the same guarantees
+   // that TR_J9SharedCache::validateClassChain() does, which is what TR_J9SharedCache::validateClassChain()
+   // uses to verify that the given clazz matches chainData. Thus we only have to check that that cached J9Class
+   // is equal to the one we are trying to validate.
+   TR::Compilation *comp = TR::compInfoPT->getCompilation();
+   bool wasReset = false;
+   auto ramClass = _deserializer->classFromOffset(chainData[1], comp, wasReset);
+   if (wasReset)
+      comp->failCompilation<J9::AOTDeserializerReset>(
+         "Deserializer reset during relocation of method %s", comp->signature());
+   TR_ASSERT_FATAL(ramClass, "RAM class for offset %zu ID %zu type %zu could not be found",
+                   chainData[1],
+                   JITServerNoSCCAOTDeserializer::offsetId(chainData[1]),
+                   JITServerNoSCCAOTDeserializer::offsetType(chainData[1]));
+   return ramClass == clazz;
+   }
+
+TR_OpaqueClassBlock *
+TR_J9DeserializerSharedCache::lookupClassFromChainAndLoader(uintptr_t *chainData, void *classLoader)
+   {
+   // The base TR_J9SharedCache::lookupClassFromChainAndLoader(), when given a class chain and class loader, will look in the loader
+   // a J9Class correspoding to the first class in the chain. If one could be found, it then verifies that the class matches the cached version.
+   // We do not need to perform that checking here, because during deserialization we will have already resolved the first class in the chain to
+   // a J9Class and verified that it matches. Thus we can simply return that cached first J9Class.
+   TR::Compilation *comp = TR::compInfoPT->getCompilation();
+   bool wasReset = false;
+   auto clazz = _deserializer->classFromOffset(chainData[1], comp, wasReset);
+   if (wasReset)
+      comp->failCompilation<J9::AOTDeserializerReset>(
+         "Deserializer reset during relocation of method %s", comp->signature());
+   TR_ASSERT_FATAL(clazz, "Class for offset %zu could not be found",
+                   chainData[1],
+                   JITServerNoSCCAOTDeserializer::offsetId(chainData[1]),
+                   JITServerNoSCCAOTDeserializer::offsetType(chainData[1]));
+   return reinterpret_cast<TR_OpaqueClassBlock *>(clazz);
```

Comment:
Here, we can simply return the cached RAM class for the first entry in the chain and ignore the `classLoader` entirely, because the deserializer will have already:

1. Looked up a candidate RAM class using its associated class loader record and cached it (so looking up another candidate in the class loader would be redundant - we'd just get the same class back as what was cached).
2. Verified that its class chain matches what was recorded at compilation (so no need to call something like `classMatchesCachedVersion`.